### DISCUSSION TITLE
Stepper Disable to Enable Timing Problem

### DISCUSSION THREAD

**Discussion by digidiver11** [grblHAL/core#299] [General]
Hello,

this is my first request or soagr an important note.
First of all I would like to thank everyone involved in this great software.
I have some questions and an important note with the request to add this in the software if necessary.

First my questions:

1. i noticed that the GRBL version 1.1f is used and not the version 1.1h.      Why is that so ?

2. Can the board "BSMCE04U" be used with a rotary encoder for the spindle synchronization 
   for a lathe and where would it have to be connected?

   I suspect that the STM32F103 CPU will not be able to do this.
  Therefore I plan for the lathe, with a spindle synchronization, the Teensy 4.1.
  Are there other microcontrollers for the spindle synchronization?

Now my important note:

I have used GRBLHAL on a BSMCE04U board for a CNC mill.

The DM542T (Z-axis) and two DM556T (X- and Y-axes) were used as output stages. Because the function "Stepper-Enable" is very important to me, I tapped this directly from the 74HC541 Pin1 and provided it via a npn-transistor as inverter with open collector as additional connection.

This connection I have directly connected with the enable connections of the stepperdriver.
This works quite well except for the sometimes strong step losses when changing from Stepper Disable to Enable.

I looked at the signals with a storage oscilloscope. The signals have a ralatively good timing despite the relatively simple hardware extension for the stepper enable pin.

Unfortunately, as is now common with the Chinese, the labeling has been removed from the stepper output stage, so it took me some searching to find the stepper IC. It is an Allegro 4989 which has a twake of 1ms. This time is also needed after the above mentioned change from Disable to Enable.

The explanation is relatively simple, this time is needed for the charging process of the bootstrap capacitors !!!.

All other stepper output stages are running fine with my invalidation. As other test stepper output stages I have used TB6600 TB6560 self-made without bootstrap capacitor. Only the output stages with Bootstrap-C make problems.

My request:
Can you please include a "pause time of about 1.5 to 2ms" after the change Disable-Enable (pin StepperEN). I.e. when the change has taken place, that the stepper pulses are output to the stepper output stages only after the delay of 1 to 1.5ms.

Many thanks in advance.

**Comment by terjeio** [comment]
> i noticed that the GRBL version 1.1f is used and not the version 1.1h. Why is that so ?

Lazyness? IMO the Grbl version number has lost its meaning now that there are several ports available with different functionality. I use the build date to determine controller functionality when I detect grblHAL firmware in [ioSender](https://github.com/terjeio/ioSender).

> Can the board "BSMCE04U" be used with a rotary encoder for the spindle synchronization

Not without adding code support for it. I have no intention of doing so for MCUs without a FPU and the STM32F103 falls in that category.

> Can you please include a "pause time of about 1.5 to 2ms" after the change Disable-Enable (pin StepperEN). I.e. when the change has taken place

There is a pause time implemented already - but somewhat randomly across drivers. For the F103 it is set in [this function](https://github.com/grblHAL/STM32F1xx/blob/1ef16aa90a7bc0c5b661d450fc660321bbced5c1/Src/driver.c#L345-L355).
Try with changing it to:

```
static void stepperWakeUp (void)
{
    stepperEnable((axes_signals_t){AXES_BITMASK});

    STEPPER_TIMER->ARR = hal.f_step_timer / 500; // ~2ms delay to allow drivers time to wake up
    STEPPER_TIMER->EGR = TIM_EGR_UG;
    STEPPER_TIMER->SR = ~TIM_SR_UIF;
    STEPPER_TIMER->CR1 |= TIM_CR1_CEN;
}
```
There will be an additional delay before the first step pulse is output - dependent on acceleration settings.

**Comment by terjeio** [comment]
> Are there other microcontrollers for the spindle synchronization?

MSP432P401R and STM32F4xx - STM32F7xx is a candidate (no supporting board maps as of now). See the [drivers overview](https://github.com/grblHAL/drivers).

**Comment by digidiver11** [comment]
Hello terjeio,

thank you for your quick answers.

To 1: Did I understand correctly that the delay for the StepperEN is already included, or do I have to insert the code lines from you if needed ?

To 2. with the spindle synchronization and the MCUs with FPU I understand well. The statement that this is only provided for these MCUs is okay.  How can I activate the spindle synchronization ?

With your ingenious WebBuilder I have found no setting for it.
If necessary I would want to compile the appropriate firmware for the Teensy 4.1 or the STM32F4xx or STM32F7xx with Arduino.
Are there any special settings or procedures, what must be considered?

Many thanks in advance.

**Reply by terjeio** [reply]
1. you have to insert the code if needed until I commit the change - I have to modify and check all drivers first... There is an implicit delay already - dependent on the acceleration used for the initial move. If you have high acceleration settings (> 500?) this can be less than 1ms.

2. Since spindle sync is work in progress I have made it a bit cumbersome. A precondition is that the board map defines the neccesary pins, [an example](https://github.com/grblHAL/STM32F4xx/blob/cbc8936502248c09a707facf07714512fc5b5b7b/Inc/st_morpho_map.h#L145-L153), and spindle sync has to be enabled in [my_machine.h](https://github.com/grblHAL/STM32F4xx/blob/cbc8936502248c09a707facf07714512fc5b5b7b/Inc/my_machine.h#LL85C1-L85C1) as well. Note that the spindle pulse input has to be a pin supported for that, it must be "routed" as a clock input to a timer (this means it can be prescaled to reduce interrupt load for high PPR encoders). 

> With your ingenious WebBuilder I have found no setting for it.

Correct - adding it is on my todo list.

>  would want to compile the appropriate firmware for the Teensy 4.1 or the STM32F4xx or STM32F7xx with Arduino.

FYI the STM32 drivers cannot be compiled with the Arduino IDE. Either STM32CubeIDE or PlatformIO must be used.

> Are there any special settings or procedures, what must be considered?

See 2 above. There are numerous settings that has to be set for spindle sync to work, among them settings for tuning the PID loop. Basic settings are $38 (encoder PPR) and $340 (spindle at speed tolerance).
